napari/components/viewer_model.py

```diff
@@ -995,6 +995,8 @@ def _add_layers_with_plugins(
             )
             # actually add the layer
             new = self._add_layer_from_data(*_data)
+            # store the source of the layer in metadata, if not present yet
+            new.metadata.setdefault('source', filename)
```

Comment:
ah right... your tests are failing because `_add_layer_from_data` can optionally return a list of layers.  That's an unfortunate design (it has to do with the fact that `add_image` can also return a list of images), but it needs to be addressed:
```python
if isinstance(new, list):
    for layer in new:
        layer.metadata.setdefault('source', filename)
else:
    new.metadata.setdefault('source', filename)
```